File: project_tools/research_c46_unpacked_permissions_debugger.mjs
#!/usr/bin/env node
import assert from 'node:assert/strict';
import crypto from 'node:crypto';
import fs from 'node:fs';
import http from 'node:http';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import puppeteer from 'puppeteer';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(HERE, '..');
const MANIFEST = JSON.parse(fs.readFileSync(path.join(ROOT, 'manifest.json'), 'utf8'));
const POPUP = fs.readFileSync(path.join(ROOT, 'popup.js'), 'utf8');
const WORKER = fs.readFileSync(path.join(ROOT, 'service-worker.js'), 'utf8');
const FRAME_AGENT = fs.readFileSync(path.join(ROOT, 'frame-agent.js'), 'utf8');

const FIXTURE_TITLE = 'C46 real unpacked debugger fixture';
const FIXTURE_MARKER = 'C46_REAL_UNPACKED_DEBUGGER_MARKER_4A7F';

function sourceChecks() {
  const permissions = new Set(MANIFEST.permissions || []);
  const optionalHosts = new Set(MANIFEST.optional_host_permissions || []);
  const backupCall = POPUP.indexOf('loadBackupStatus();');
  const activeTabDef = POPUP.indexOf('async function getActiveSourceTab');
  const grantStart = POPUP.indexOf("grantFrameAccessButton?.addEventListener('click'");
  const grantBlock = grantStart >= 0 ? POPUP.slice(grantStart, grantStart + 12000) : '';
  const discovery = grantBlock.indexOf('await collectCrossOriginFrameOrigins');
  const permissionRequest = grantBlock.indexOf('chrome.permissions.request');
  const timeoutMatch = POPUP.match(/POPUP_EXTENSION_API_TIMEOUT_MS\s*=\s*([0-9_]+)/);
  const timeoutMs = timeoutMatch ? Number(timeoutMatch[1].replaceAll('_', '')) : null;

  const checks = {
    manifestDebuggerRequired: permissions.has('debugger'),
    manifestOptionalHttpHost: optionalHosts.has('http://*/*'),
    manifestOptionalHttpsHost: optionalHosts.has('https://*/*'),
    manifestIncognitoKeyPresent: Object.hasOwn(MANIFEST, 'incognito'),
    popupBackupStatusBeforeActiveTabClassification: backupCall >= 0 && activeTabDef >= 0 && backupCall < activeTabDef,
    popupPermissionRequestAfterAsyncDiscovery: discovery >= 0 && permissionRequest >= 0 && discovery < permissionRequest,
    popupCallsChromePermissionsRequest: POPUP.includes('chrome.permissions.request'),
    popupExtensionApiTimeoutMs: timeoutMs,
    contentSenderIncognitoGuardPresent: WORKER.includes('sender?.tab?.incognito'),
    permissionsOnRemovedListenerPresent: WORKER.includes('chrome.permissions.onRemoved'),
    frameAgentPermissionRecheckPresent: WORKER.includes('frameAgentHasGrantedHostPermission'),
    frameAgentMemoryRegistryPresent: WORKER.includes('frameAgentsByTab = new Map()'),
    frameAgentReinjectReregistersWithoutReset:
      FRAME_AGENT.includes('__WEBCLIP_FRAME_AGENT_LOADED__') &&
      FRAME_AGENT.includes('WEBCLIP_FRAME_AGENT_REGISTER') &&
      FRAME_AGENT.indexOf('__WEBCLIP_FRAME_AGENT_LOADED__') < FRAME_AGENT.indexOf('const state ='),
    debuggerActiveRegistryPresent: WORKER.includes('debuggerActiveTabs = new Set()'),
    debuggerLateAttachCleanupPresent: WORKER.includes('debuggerLateAttachCleanupByTab = new Map()'),
    debuggerPendingDetachPresent: WORKER.includes('debuggerPendingDetachByTab = new Map()'),
    debuggerActualSettlementTrackingPresent: WORKER.includes('debuggerPendingActualSettlements = new Set()'),
    boundedDebuggerAttachPresent: WORKER.includes('async function attachDebuggerBounded'),
    boundedDebuggerDetachPresent: WORKER.includes('async function detachDebuggerBounded'),
    physicalDebuggerPdfPathPresent: WORKER.includes("chrome.debugger.sendCommand(debuggee, 'Page.printToPDF'"),
    debuggerFinallyDropsActiveTab: WORKER.includes('debuggerActiveTabs.delete(tabId)'),
    debuggerOnDetachHandlingPresent: WORKER.includes('debugger.onDetach'),
  };

  for (const key of [
    'manifestDebuggerRequired', 'manifestOptionalHttpHost', 'manifestOptionalHttpsHost',
    'popupBackupStatusBeforeActiveTabClassification', 'popupPermissionRequestAfterAsyncDiscovery',
    'popupCallsChromePermissionsRequest', 'contentSenderIncognitoGuardPresent',
    'frameAgentPermissionRecheckPresent', 'frameAgentMemoryRegistryPresent',
    'frameAgentReinjectReregistersWithoutReset', 'debuggerActiveRegistryPresent',
    'debuggerLateAttachCleanupPresent', 'debuggerPendingDetachPresent',
    'debuggerActualSettlementTrackingPresent', 'boundedDebuggerAttachPresent',
    'boundedDebuggerDetachPresent', 'physicalDebuggerPdfPathPresent', 'debuggerFinallyDropsActiveTab',
  ]) assert.equal(checks[key], true, `${key} source guard`);
  assert.equal(checks.manifestIncognitoKeyPresent, false);
  assert.equal(checks.permissionsOnRemovedListenerPresent, false);
  assert.equal(checks.debuggerOnDetachHandlingPresent, false);
  assert.equal(checks.popupExtensionApiTimeoutMs, 10000);
  return checks;
}

async function startFixture() {
  let reportResolve;
  let reportReject;
  const reportPromise = new Promise((resolve, reject) => {
    reportResolve = resolve;
    reportReject = reject;
  });
  const server = http.createServer((req, res) => {
    const url = new URL(req.url || '/', 'http://127.0.0.1');
    if (url.pathname === '/fixture') {
      const body = `<!doctype html><meta charset="utf-8"><title>${FIXTURE_TITLE}</title><main><h1>${FIXTURE_MARKER}</h1><p>Production debugger path fixture.</p></main>`;
      res.writeHead(200, { 'content-type': 'text/html; charset=utf-8', 'cache-control': 'no-store' });
      res.end(body);
      return;
    }
    if (url.pathname === '/report') {
      try {
        reportResolve(JSON.parse(url.searchParams.get('data') || '{}'));
        res.writeHead(200, { 'content-type': 'text/plain', 'cache-control': 'no-store' });
        res.end('ok');
      } catch (error) {
        reportReject(error);
        res.writeHead(400); res.end('bad report');
      }
      return;
    }
    res.writeHead(404); res.end('not found');
  });
  await new Promise((resolve, reject) => {
    server.once('error', reject);
    server.listen(0, '127.0.0.1', resolve);
  });
  const address = server.address();
  const base = `http://127.0.0.1:${address.port}`;
  return {
    server,
    fixtureUrl: `${base}/fixture`,
    reportUrl: `${base}/report`,
    originPattern: `${base}/*`,
    reportPromise,
  };
}

function testDriver({ fixtureUrl, reportUrl, originPattern }) {
  return `\n\n// C46 test-only sidecar appended only to a temporary unpacked copy.\n(async () => {\n  const out = { ok: false, stage: 'startup' };\n  let tab = null;\n  const report = async () => {\n    const data = encodeURIComponent(JSON.stringify(out));\n    try {\n      if (tab?.id) await chrome.tabs.update(tab.id, { url: ${JSON.stringify(reportUrl)} + '?data=' + data });\n      else await chrome.tabs.create({ url: ${JSON.stringify(reportUrl)} + '?data=' + data, active: false });\n    } catch (_) {}\n  };\n  const waitForLoaded = async (tabId) => {\n    const current = await chrome.tabs.get(tabId);\n    if (current.status === 'complete') return;\n    await new Promise((resolve, reject) => {\n      const timer = setTimeout(() => { chrome.tabs.onUpdated.removeListener(listener); reject(new Error('fixture-load-timeout')); }, 15000);\n      const listener = (updatedTabId, changeInfo) => {\n        if (updatedTabId === tabId && changeInfo.status === 'complete') {\n          clearTimeout(timer); chrome.tabs.onUpdated.removeListener(listener); resolve();\n        }\n      };\n      chrome.tabs.onUpdated.addListener(listener);\n    });\n  };\n  try {\n    await new Promise(resolve => setTimeout(resolve, 1500));\n    out.stage = 'incognito-access';\n    out.incognitoAllowed = await chrome.extension.isAllowedIncognitoAccess();\n    out.stage = 'optional-permission-state';\n    out.optionalOriginGrantedInitially = await chrome.permissions.contains({ origins: [${JSON.stringify(originPattern)}] });\n    out.stage = 'tab-create';\n    tab = await chrome.tabs.create({ url: ${JSON.stringify(fixtureUrl)}, active: false });\n    out.tabId = tab.id;\n    out.stage = 'tab-load';\n    await waitForLoaded(tab.id);\n    out.fixtureTitle = (await chrome.tabs.get(tab.id)).title || '';\n    out.stage = 'debugger-target-before';\n    const before = (await chrome.debugger.getTargets()).find(x => x.tabId === tab.id);\n    out.debuggerAttachedBefore = Boolean(before?.attached);\n    out.stage = 'production-generate-pdf';\n    const pdfBlob = await generatePdfBlob(tab.id);\n    out.stage = 'pdf-digest';\n    const bytes = new Uint8Array(await pdfBlob.arrayBuffer());\n    const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', bytes));\n    out.pdfBytes = bytes.byteLength;\n    out.pdfSha256 = [...digest].map(x => x.toString(16).padStart(2, '0')).join('');\n    out.pdfHeaderHex = [...bytes.subarray(0, 8)].map(x => x.toString(16).padStart(2, '0')).join('');\n    out.stage = 'debugger-target-after';\n    const after = (await chrome.debugger.getTargets()).find(x => x.tabId === tab.id);\n    out.debuggerAttachedAfter = Boolean(after?.attached);\n    out.debuggerActiveSetAfter = debuggerActiveTabs.has(tab.id);\n    out.debuggerLateAttachCleanupAfter = debuggerLateAttachCleanupByTab.has(tab.id);\n    out.debuggerPendingDetachAfter = debuggerPendingDetachByTab.has(tab.id);\n    out.debuggerPendingActualSettlementCountAfter = debuggerPendingActualSettlements.size;\n    out.stage = 'done';\n    out.ok = true;\n  } catch (error) {\n    out.error = String(error?.stack || error?.message || error);\n  }\n  await report();\n})();\n`;
}

function makeTemporaryExtension(fixture) {
  const tempRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'webclip-c46-'));
  const extensionRoot = path.join(tempRoot, 'extension');
  fs.cpSync(ROOT, extensionRoot, {
    recursive: true,
    filter(source) {
      const rel = path.relative(ROOT, source);
      if (!rel) return true;
      const first = rel.split(path.sep)[0];
      return !['.git', 'node_modules'].includes(first);
    },
  });
  const workerPath = path.join(extensionRoot, 'service-worker.js');
  fs.appendFileSync(workerPath, testDriver(fixture), 'utf8');
  return { tempRoot, extensionRoot };
}

async function runRealUnpacked(fixture) {
  const temporary = makeTemporaryExtension(fixture);
  let browser = null;
  let browserProcess = null;
  try {
    browser = await puppeteer.launch({
      headless: false,
      pipe: true,
      enableExtensions: [temporary.extensionRoot],
      args: ['--no-sandbox', '--disable-gpu', '--disable-dev-shm-usage', '--no-first-run', '--no-default-browser-check'],
    });
    browserProcess = browser.process();
    const browserVersion = await browser.version();
    const workerTarget = await browser.waitForTarget(
      target => target.type() === 'service_worker' && target.url().endsWith('/service-worker.js'),
      { timeout: 30000 },
    );
    const extensionId = new URL(workerTarget.url()).host;

    browser.disconnect();
    browser = null;

    const result = await Promise.race([
      fixture.reportPromise,
      new Promise((_, reject) => setTimeout(() => reject(new Error('C46 report timeout after Puppeteer disconnect')), 30000)),
    ]);

    assert.equal(result.ok, true, `browser stage failed: ${JSON.stringify(result)}`);
    assert.equal(result.optionalOriginGrantedInitially, false, 'optional host is not silently granted');
    assert.equal(result.debuggerAttachedBefore, false, 'debugger starts detached after Puppeteer disconnect');
    assert.equal(result.debuggerAttachedAfter, false, 'debugger finishes detached');
    assert.equal(result.debuggerActiveSetAfter, false, 'active registry cleaned');
    assert.equal(result.debuggerLateAttachCleanupAfter, false, 'late-attach registry cleaned');
    assert.equal(result.debuggerPendingDetachAfter, false, 'pending-detach registry cleaned');
    assert.equal(result.debuggerPendingActualSettlementCountAfter, 0, 'all debugger API promises settled');
    assert.equal(result.fixtureTitle, FIXTURE_TITLE, 'correct fixture tab was printed');
    assert(Number(result.pdfBytes) > 1000, 'physical PDF bytes');
    assert.match(String(result.pdfSha256 || ''), /^[0-9a-f]{64}$/);
    assert(String(result.pdfHeaderHex || '').startsWith('255044462d'), 'physical PDF starts with %PDF-');

    return { browserVersion, extensionId, ...result };
  } finally {
    if (browser) {
      try { await browser.close(); } catch (_) {}
    }
    if (browserProcess && browserProcess.exitCode == null) {
      try { browserProcess.kill('SIGTERM'); } catch (_) {}
    }
    fs.rmSync(temporary.tempRoot, { recursive: true, force: true });
  }
}

const source = sourceChecks();
const fixture = await startFixture();
try {
  const realUnpacked = await runRealUnpacked(fixture);
  const result = {
    sourceBaseline: process.env.C46_SOURCE_BASELINE || '',
    source,
    realUnpacked,
    classification: 'L4 real-unpacked positive production chrome.debugger/Page.printToPDF lifecycle plus exact-source private-context/permission/revoke findings; interactive incognito toggle, native permission prompt, revoke/regrant and worker/browser restart remain L5 open',
  };
  const canonical = JSON.stringify(result);
  result.resultSha256 = crypto.createHash('sha256').update(canonical).digest('hex');
  console.log(`C46_RESULT_JSON=${JSON.stringify(result)}`);
} finally {
  await new Promise(resolve => fixture.server.close(resolve));
}

File: project_tools/research_c46_unpacked_permissions_debugger.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import hashlib
import json
import os
import pathlib
import re
import shutil
import subprocess
import tempfile
import threading
import urllib.parse
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

ROOT = pathlib.Path(__file__).resolve().parents[1]
MANIFEST = json.loads((ROOT / 'manifest.json').read_text(encoding='utf-8'))
POPUP = (ROOT / 'popup.js').read_text(encoding='utf-8')
WORKER = (ROOT / 'service-worker.js').read_text(encoding='utf-8')
FRAME_AGENT = (ROOT / 'frame-agent.js').read_text(encoding='utf-8')
CHROME_DEFAULT = os.environ.get('CHROMIUM_BIN', shutil.which('google-chrome') or shutil.which('chromium') or '')

FIXTURE_TITLE = 'C46 real unpacked debugger fixture'
FIXTURE_MARKER = 'C46_REAL_UNPACKED_DEBUGGER_MARKER_4A7F'


def source_checks() -> dict:
    permissions = set(MANIFEST.get('permissions') or [])
    optional_hosts = set(MANIFEST.get('optional_host_permissions') or [])

    backup_call_idx = POPUP.find('loadBackupStatus();')
    active_tab_def_idx = POPUP.find('async function getActiveSourceTab')

    grant_idx = POPUP.find("grantFrameAccessButton?.addEventListener('click'")
    grant_block = POPUP[grant_idx:grant_idx + 12000] if grant_idx >= 0 else ''
    discovery_idx = grant_block.find('await collectCrossOriginFrameOrigins')
    permission_request_idx = grant_block.find('chrome.permissions.request')

    popup_timeout_match = re.search(r'POPUP_EXTENSION_API_TIMEOUT_MS\s*=\s*([0-9_]+)', POPUP)
    popup_timeout = int(popup_timeout_match.group(1).replace('_', '')) if popup_timeout_match else None

    checks = {
        'manifestDebuggerRequired': 'debugger' in permissions,
        'manifestOptionalHttpHost': 'http://*/*' in optional_hosts,
        'manifestOptionalHttpsHost': 'https://*/*' in optional_hosts,
        'manifestIncognitoKeyPresent': 'incognito' in MANIFEST,
        'popupBackupStatusBeforeActiveTabClassification': backup_call_idx >= 0 and active_tab_def_idx >= 0 and backup_call_idx < active_tab_def_idx,
        'popupPermissionRequestAfterAsyncDiscovery': discovery_idx >= 0 and permission_request_idx >= 0 and discovery_idx < permission_request_idx,
        'popupCallsChromePermissionsRequest': 'chrome.permissions.request' in POPUP,
        'popupExtensionApiTimeoutMs': popup_timeout,
        'contentSenderIncognitoGuardPresent': 'sender?.tab?.incognito' in WORKER,
        'permissionsOnRemovedListenerPresent': 'chrome.permissions.onRemoved' in WORKER,
        'frameAgentPermissionRecheckPresent': 'frameAgentHasGrantedHostPermission' in WORKER,
        'frameAgentMemoryRegistryPresent': 'frameAgentsByTab = new Map()' in WORKER,
        'frameAgentReinjectReregistersWithoutReset': '__WEBCLIP_FRAME_AGENT_LOADED__' in FRAME_AGENT and 'WEBCLIP_FRAME_AGENT_REGISTER' in FRAME_AGENT and FRAME_AGENT.find('__WEBCLIP_FRAME_AGENT_LOADED__') < FRAME_AGENT.find('const state ='),
        'debuggerActiveRegistryPresent': 'debuggerActiveTabs = new Set()' in WORKER,
        'debuggerLateAttachCleanupPresent': 'debuggerLateAttachCleanupByTab = new Map()' in WORKER,
        'debuggerPendingDetachPresent': 'debuggerPendingDetachByTab = new Map()' in WORKER,
        'debuggerActualSettlementTrackingPresent': 'debuggerPendingActualSettlements = new Set()' in WORKER,
        'boundedDebuggerAttachPresent': 'async function attachDebuggerBounded' in WORKER,
        'boundedDebuggerDetachPresent': 'async function detachDebuggerBounded' in WORKER,
        'physicalDebuggerPdfPathPresent': "chrome.debugger.sendCommand(debuggee, 'Page.printToPDF'" in WORKER,
        'debuggerFinallyDropsActiveTab': 'debuggerActiveTabs.delete(tabId)' in WORKER,
        'debuggerOnDetachHandlingPresent': 'debugger.onDetach' in WORKER,
    }

    required_true = [
        'manifestDebuggerRequired',
        'manifestOptionalHttpHost',
        'manifestOptionalHttpsHost',
        'popupBackupStatusBeforeActiveTabClassification',
        'popupPermissionRequestAfterAsyncDiscovery',
        'popupCallsChromePermissionsRequest',
        'contentSenderIncognitoGuardPresent',
        'frameAgentPermissionRecheckPresent',
        'frameAgentMemoryRegistryPresent',
        'frameAgentReinjectReregistersWithoutReset',
        'debuggerActiveRegistryPresent',
        'debuggerLateAttachCleanupPresent',
        'debuggerPendingDetachPresent',
        'debuggerActualSettlementTrackingPresent',
        'boundedDebuggerAttachPresent',
        'boundedDebuggerDetachPresent',
        'physicalDebuggerPdfPathPresent',
        'debuggerFinallyDropsActiveTab',
    ]
    for key in required_true:
        assert checks[key] is True, (key, checks)
    assert checks['manifestIncognitoKeyPresent'] is False, checks
    assert checks['permissionsOnRemovedListenerPresent'] is False, checks
    assert checks['debuggerOnDetachHandlingPresent'] is False, checks
    assert checks['popupExtensionApiTimeoutMs'] == 10_000, checks
    return checks


class C46Handler(BaseHTTPRequestHandler):
    report_event = threading.Event()
    report_payload: dict | None = None

    def do_GET(self):
        parsed = urllib.parse.urlparse(self.path)
        if parsed.path == '/fixture':
            body = (
                '<!doctype html><meta charset="utf-8">'
                f'<title>{FIXTURE_TITLE}</title>'
                f'<main><h1>{FIXTURE_MARKER}</h1><p>Production debugger path fixture.</p></main>'
            ).encode('utf-8')
            self.send_response(200)
            self.send_header('Content-Type', 'text/html; charset=utf-8')
            self.send_header('Cache-Control', 'no-store')
            self.send_header('Content-Length', str(len(body)))
            self.end_headers()
            self.wfile.write(body)
            return
        if parsed.path == '/report':
            query = urllib.parse.parse_qs(parsed.query)
            raw = (query.get('data') or ['{}'])[0]
            try:
                type(self).report_payload = json.loads(raw)
            except Exception as exc:
                type(self).report_payload = {'ok': False, 'error': f'report-json:{exc}', 'raw': raw[:1000]}
            type(self).report_event.set()
            body = b'ok'
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain')
            self.send_header('Content-Length', str(len(body)))
            self.end_headers()
            self.wfile.write(body)
            return
        self.send_response(404)
        self.end_headers()

    def log_message(self, *_):
        pass


def browser_version(chrome: str) -> str:
    try:
        return subprocess.check_output([chrome, '--version'], text=True, stderr=subprocess.STDOUT, timeout=10).strip()
    except Exception:
        return ''


def js_driver(fixture_url: str, report_url: str) -> str:
    origin = urllib.parse.urlsplit(fixture_url)
    pattern = f'{origin.scheme}://{origin.hostname}:{origin.port}/*'
    return f'''

// C46 test-only driver appended to a temporary unpacked copy by the durable Python harness.
// It invokes production generatePdfBlob() unchanged and reports through a normal tab navigation.
(async () => {{
  const fixtureUrl = {json.dumps(fixture_url)};
  const reportBase = {json.dumps(report_url)};
  const optionalOrigin = {json.dumps(pattern)};
  const result = {{ ok: false, extensionId: chrome.runtime.id }};
  const report = async () => {{
    const data = encodeURIComponent(JSON.stringify(result));
    try {{
      const tabs = await chrome.tabs.query({{}});
      const tab = tabs.find((x) => Number(x.id) === Number(result.tabId)) || tabs[0];
      if (tab?.id) await chrome.tabs.update(tab.id, {{ url: `${{reportBase}}?data=${{data}}` }});
      else await chrome.tabs.create({{ url: `${{reportBase}}?data=${{data}}`, active: false }});
    }} catch (_) {{}}
  }};
  try {{
    result.incognitoAllowed = await chrome.extension.isAllowedIncognitoAccess();
    result.optionalOriginGrantedInitially = await chrome.permissions.contains({{ origins: [optionalOrigin] }});
    const tab = await chrome.tabs.create({{ url: fixtureUrl, active: false }});
    result.tabId = tab.id;
    await new Promise((resolve, reject) => {{
      const timer = setTimeout(() => {{ chrome.tabs.onUpdated.removeListener(listener); reject(new Error('fixture-load-timeout')); }}, 15000);
      const listener = (tabId, info) => {{
        if (tabId === tab.id && info.status === 'complete') {{ clearTimeout(timer); chrome.tabs.onUpdated.removeListener(listener); resolve(); }}
      }};
      chrome.tabs.onUpdated.addListener(listener);
    }});
    const before = (await chrome.debugger.getTargets()).find((x) => x.tabId === tab.id);
    result.debuggerAttachedBefore = Boolean(before?.attached);

    const pdfBlob = await generatePdfBlob(tab.id);
    const bytes = new Uint8Array(await pdfBlob.arrayBuffer());
    const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', bytes));
    result.pdfBytes = bytes.byteLength;
    result.pdfSha256 = [...digest].map((x) => x.toString(16).padStart(2, '0')).join('');

    const after = (await chrome.debugger.getTargets()).find((x) => x.tabId === tab.id);
    result.debuggerAttachedAfter = Boolean(after?.attached);
    result.debuggerActiveSetAfter = debuggerActiveTabs.has(tab.id);
    result.debuggerLateAttachCleanupAfter = debuggerLateAttachCleanupByTab.has(tab.id);
    result.debuggerPendingDetachAfter = debuggerPendingDetachByTab.has(tab.id);
    result.debuggerPendingActualSettlementCountAfter = debuggerPendingActualSettlements.size;
    result.fixtureTitle = (await chrome.tabs.get(tab.id)).title || '';
    result.ok = true;
  }} catch (error) {{
    result.error = String(error?.stack || error?.message || error);
  }}
  await report();
}})();
'''


def run_real_unpacked(chrome: str, timeout_seconds: int = 45) -> dict:
    C46Handler.report_event.clear()
    C46Handler.report_payload = None
    server = ThreadingHTTPServer(('127.0.0.1', 0), C46Handler)
    threading.Thread(target=server.serve_forever, daemon=True).start()
    port = server.server_address[1]
    fixture_url = f'http://127.0.0.1:{port}/fixture'
    report_url = f'http://127.0.0.1:{port}/report'

    temp_root = pathlib.Path(tempfile.mkdtemp(prefix='webclip-c46-'))
    ext_root = temp_root / 'extension'
    profile = temp_root / 'profile'
    try:
        shutil.copytree(ROOT, ext_root, ignore=shutil.ignore_patterns('.git', '__pycache__', '*.pyc'))
        worker_path = ext_root / 'service-worker.js'
        worker_path.write_text(WORKER + js_driver(fixture_url, report_url), encoding='utf-8')

        cmd = [
            chrome,
            '--headless=new',
            '--no-sandbox',
            '--disable-gpu',
            '--disable-dev-shm-usage',
            '--no-first-run',
            '--no-default-browser-check',
            '--disable-background-networking',
            f'--user-data-dir={profile}',
            f'--disable-extensions-except={ext_root}',
            f'--load-extension={ext_root}',
            'about:blank',
        ]
        proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        try:
            if not C46Handler.report_event.wait(timeout_seconds):
                stderr = ''
                if proc.poll() is not None:
                    try:
                        stderr = (proc.stderr.read() if proc.stderr else '')[-4000:]
                    except Exception:
                        pass
                raise AssertionError(f'C46 real-unpacked report timeout; chrome_rc={proc.poll()} stderr={stderr}')
            payload = C46Handler.report_payload or {}
        finally:
            proc.terminate()
            try:
                proc.wait(timeout=8)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=5)

        assert payload.get('ok') is True, payload
        assert payload.get('debuggerAttachedBefore') is False, payload
        assert payload.get('debuggerAttachedAfter') is False, payload
        assert payload.get('debuggerActiveSetAfter') is False, payload
        assert payload.get('debuggerLateAttachCleanupAfter') is False, payload
        assert payload.get('debuggerPendingDetachAfter') is False, payload
        assert int(payload.get('debuggerPendingActualSettlementCountAfter') or 0) == 0, payload
        assert int(payload.get('pdfBytes') or 0) > 1000, payload
        assert re.fullmatch(r'[0-9a-f]{64}', str(payload.get('pdfSha256') or '')), payload
        assert payload.get('optionalOriginGrantedInitially') is False, payload
        return payload
    finally:
        server.shutdown()
        server.server_close()
        shutil.rmtree(temp_root, ignore_errors=True)


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument('--chrome', default=CHROME_DEFAULT)
    ap.add_argument('--source-only', action='store_true')
    ap.add_argument('--source-baseline', default=os.environ.get('C46_SOURCE_BASELINE', ''))
    args = ap.parse_args()

    source = source_checks()
    browser = None
    if not args.source_only:
        if not args.chrome:
            raise SystemExit('Chrome unavailable; use --source-only for deterministic source checks')
        browser = run_real_unpacked(args.chrome)

    result = {
        'sourceBaseline': args.source_baseline,
        'browserVersion': browser_version(args.chrome) if args.chrome else '',
        'source': source,
        'realUnpacked': browser,
        'classification': (
            'L4 real-unpacked positive debugger/PDF lifecycle plus exact-source privacy/permission/revoke findings; '
            'interactive incognito toggle, native permission prompt, revoke/regrant UI and worker/browser restart remain L5 open'
            if browser else
            'L3 exact-source privacy/permission/revoke/debugger lifecycle evidence only'
        ),
    }
    canonical = json.dumps(result, ensure_ascii=False, sort_keys=True, separators=(',', ':')).encode('utf-8')
    result['resultSha256'] = hashlib.sha256(canonical).hexdigest()
    print('C46_RESULT_JSON=' + json.dumps(result, ensure_ascii=False, sort_keys=True, separators=(',', ':')), flush=True)


if __name__ == '__main__':
    main()
